### ISSUE TITLE
Black textures on low-end systems after minutes of playthrough

### ISSUE THREAD

**Issue by slash0042** [gonetz/GLideN64#2269]
Hi, Mupen64Plus Next Switch user here.
This has been an issue I was experiencing since N64 emulation became a thing on Switch, the behavior is this one showed on this clip: https://streamable.com/u51oky

I commented this issue to @m4xw and he said since the assets/textures are never unloaded it reaches to the Switch's bandwidth limits, this doesn't seem to be an issue on PC (unless we are talking about a really old one). As you can see, I'm using an ESRGAN enhanced texture pack, but descripted on #2152 , this can happen with vanilla textures, although it is more ease to trigger it with hires texture packs. On Switch we have 4GB of RAM (only 3 of em are usable), so if you need help on testing to isolate the issue feel free to ask for.

**Comment by m4xw** [comment]
A solution to this issue (with native textures) was lower max texture cache size to 4000 entries (down from 8000)
On that note, I think the logic to should be reworked to be based on the size of the entries, not the amount.
as for high res textures, these issues get triggered because AFAIK the High res textures never get unloaded from ram after being lazy loaded.
BTW gonetz what do you think about using normal textures and replace them on the fly as high res textures get loaded (flycast does that for example)
could get rid of loading Stutter even more.

**Comment by gonetz** [comment]
> AFAIK the High res textures never get unloaded from ram after being lazy loaded.

Yes. It works perfect if you have enough RAM. If amount of RAM is not large enough or if texture pack is too large (as Nerrel's pack for Zelda MM) you have another option: **file storage**. In that case hd textures will not be cached in RAM. They will be loaded from hdd right into video memory. It hardly help with the issue on Switch though, since, as I understand, it has low video memory too. "lower max texture cache size to 4000 entries" may help.

**Comment by m4xw** [comment]
> 
> 
> > AFAIK the High res textures never get unloaded from ram after being lazy loaded.
> 
> Yes. It works perfect if you have enough RAM. If amount of RAM is not large enough or if texture pack is too large (as Nerrel's pack for Zelda MM) you have another option: **file storage**. In that case hd textures will not be cached in RAM. They will be loaded from hdd right into video memory. It hardly help with the issue on Switch though, since, as I understand, it has low video memory too. "lower max texture cache size to 4000 entries" may help.

This is with enhanced file storage, the problem just happens after a bit of playing now as more and more get used.

> They will be loaded from hdd right into video memory.

That's what i mean with lazy loading ;)
> "lower max texture cache size to 4000 entries" may help.

4000 is too much with bigger packs even.
Really need to use a metric that resembles a proper metric.

**Comment by slash0042** [comment]
> > AFAIK the High res textures never get unloaded from ram after being lazy loaded.
> 
> Yes. It works perfect if you have enough RAM. If amount of RAM is not large enough or if texture pack is too large (as Nerrel's pack for Zelda MM) you have another option: **file storage**. In that case hd textures will not be cached in RAM. They will be loaded from hdd right into video memory. It hardly help with the issue on Switch though, since, as I understand, it has low video memory too. "lower max texture cache size to 4000 entries" may help.

I should have mentioned that I'm using enhanced file storage + cache size is set to 4000 and as seen on the video the issue is still there, same happens with RE2 with vanilla textures even if the cache size is set to 4000.

**Comment by gonetz** [comment]
>>Really need to use a metric that resembles smth useful other than cache entries

we didn't find a safe way to find video memory size with OpenGL.
